File: test_scripts/Live Tracking for Shipments.js
describe('Live Tracking for Shipments', () => {
  const shipmentId = '54321';
  const userId = '09876';
  const trackingPage = new LiveTrackingPage();

  before(() => {
    cy.login(userId);
  });

  it('should navigate to the live tracking page', () => {
    trackingPage.navigate();
    trackingPage.verifyPageDisplayed();
  });

  it('should enter shipment ID and display tracking details', () => {
    trackingPage.enterShipmentId(shipmentId);
    trackingPage.verifyTrackingDetailsDisplayed(shipmentId);
  });

  it('should check current location of the shipment', () => {
    trackingPage.verifyCurrentLocationDisplayed();
  });

  it('should simulate a change in shipment location', () => {
    trackingPage.simulateLocationChange();
    trackingPage.verifyLocationUpdatesRealTime();
  });

  it('should verify the estimated delivery time', () => {
    trackingPage.verifyEstimatedDeliveryTime();
  });

  it('should refresh the page and verify tracking information consistency', () => {
    trackingPage.refreshPage();
    trackingPage.verifyTrackingInfoConsistency();
  });

  it('should log out and log back in, verifying tracking information availability', () => {
    cy.logout();
    cy.login(userId);
    trackingPage.verifyTrackingInfoAvailability();
  });

  it('should check notification settings for live tracking updates', () => {
    trackingPage.verifyNotificationSettings();
  });

  it('should simulate network issue and verify graceful handling', () => {
    trackingPage.simulateNetworkIssue();
    trackingPage.verifyNetworkIssueHandling();
  });

  it('should verify tracking history log', () => {
    trackingPage.verifyTrackingHistoryLog();
  });

  it('should check for error messages during location updates', () => {
    trackingPage.verifyNoErrorMessages();
  });

  it('should attempt location update from a different device', () => {
    trackingPage.verifyLocationUpdateSyncAcrossDevices();
  });

  it('should verify live tracking on a mobile device', () => {
    trackingPage.verifyTrackingConsistencyMobile();
  });

  it('should check live tracking after system reboot', () => {
    trackingPage.verifyTrackingInfoPostReboot();
  });

  it('should verify accuracy of the live tracking system', () => {
    trackingPage.verifyTrackingAccuracy();
  });
});

File: bdd/step_definitions/Live Tracking for Shipments.java
package com.cucumber.steps;

import com.framework.cucumber.TestHarness;
import com.page_objects.LiveTrackingPage;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;
import java.util.List;

public class LiveTrackingSteps extends LiveTrackingPage {

    private TestHarness testHarness = new TestHarness();

    @Given("the user is logged into the system")
    public void theUserIsLoggedIntoTheSystem() {
        loginToSystem();
    }

    @When("the user navigates to the live tracking page")
    public void theUserNavigatesToTheLiveTrackingPage() {
        navigateToLiveTrackingPage();
    }

    @Then("the live tracking page is displayed")
    public void theLiveTrackingPageIsDisplayed() {
        verifyLiveTrackingPageDisplayed();
    }

    @Given("the user is on the live tracking page")
    public void theUserIsOnTheLiveTrackingPage() {
        verifyLiveTrackingPageDisplayed();
    }

    @When("the user enters Shipment ID \"54321\" in the tracking field")
    public void theUserEntersShipmentIDInTheTrackingField() {
        enterShipmentID("54321");
    }

    @Then("live tracking details for ID \"54321\" are displayed")
    public void liveTrackingDetailsForIDAreDisplayed() {
        verifyTrackingDetailsDisplayed("54321");
    }

    @Given("live tracking details for ID \"54321\" are displayed")
    public void liveTrackingDetailsForIDAreDisplayedAgain() {
        verifyTrackingDetailsDisplayed("54321");
    }

    @When("the user checks the current location of the shipment")
    public void theUserChecksTheCurrentLocationOfTheShipment() {
        checkCurrentLocation();
    }

    @Then("the current location is displayed on the map")
    public void theCurrentLocationIsDisplayedOnTheMap() {
        verifyCurrentLocationDisplayed();
    }

    @Given("the current location of the shipment is displayed on the map")
    public void theCurrentLocationOfTheShipmentIsDisplayedOnTheMap() {
        verifyCurrentLocationDisplayed();
    }

    @When("the shipment location changes")
    public void theShipmentLocationChanges() {
        simulateLocationChange();
    }

    @Then("the location updates in real-time on the map")
    public void theLocationUpdatesInRealTimeOnTheMap() {
        verifyLocationUpdatesRealTime();
    }

    @When("the user checks the estimated delivery time")
    public void theUserChecksTheEstimatedDeliveryTime() {
        checkEstimatedDeliveryTime();
    }

    @Then("the estimated delivery time is displayed accurately")
    public void theEstimatedDeliveryTimeIsDisplayedAccurately() {
        verifyEstimatedDeliveryTime();
    }

    @When("the user refreshes the page")
    public void theUserRefreshesThePage() {
        refreshPage();
    }

    @Then("live tracking information remains consistent")
    public void liveTrackingInformationRemainsConsistent() {
        verifyTrackingConsistency();
    }

    @When("the user logs out and logs back in")
    public void theUserLogsOutAndLogsBackIn() {
        logoutAndLogin();
    }

    @Then("live tracking information remains available")
    public void liveTrackingInformationRemainsAvailable() {
        verifyTrackingAvailability();
    }

    @When("the user checks the notification settings for live tracking updates")
    public void theUserChecksTheNotificationSettingsForLiveTrackingUpdates() {
        checkNotificationSettings();
    }

    @Then("notifications are enabled for live tracking updates")
    public void notificationsAreEnabledForLiveTrackingUpdates() {
        verifyNotificationsEnabled();
    }

    @When("a network issue occurs and the system attempts to update location")
    public void aNetworkIssueOccursAndTheSystemAttemptsToUpdateLocation() {
        simulateNetworkIssue();
    }

    @Then("the system handles the issue gracefully and retries the update")
    public void theSystemHandlesTheIssueGracefullyAndRetriesTheUpdate() {
        verifyNetworkIssueHandling();
    }

    @When("the user verifies the tracking history log")
    public void theUserVerifiesTheTrackingHistoryLog() {
        verifyTrackingHistoryLog();
    }

    @Then("the log contains all location updates with timestamps")
    public void theLogContainsAllLocationUpdatesWithTimestamps() {
        verifyLogContents();
    }

    @When("the user checks for any error messages during location updates")
    public void theUserChecksForAnyErrorMessagesDuringLocationUpdates() {
        checkForErrorMessages();
    }

    @Then("no error messages are displayed")
    public void noErrorMessagesAreDisplayed() {
        verifyNoErrorMessages();
    }

    @When("the user updates location from a different device")
    public void theUserUpdatesLocationFromADifferentDevice() {
        updateLocationFromDifferentDevice();
    }

    @Then("location updates are synchronized across devices")
    public void locationUpdatesAreSynchronizedAcrossDevices() {
        verifySynchronizationAcrossDevices();
    }

    @When("the user verifies the live tracking on a mobile device")
    public void theUserVerifiesTheLiveTrackingOnAMobileDevice() {
        verifyLiveTrackingOnMobile();
    }

    @Then("tracking is consistent with the desktop view")
    public void trackingIsConsistentWithTheDesktopView() {
        verifyConsistencyWithDesktopView();
    }

    @When("the system reboots")
    public void theSystemReboots() {
        simulateSystemReboot();
    }

    @Then("tracking information remains available")
    public void trackingInformationRemainsAvailable() {
        verifyTrackingAvailabilityAfterReboot();
    }

    @When("the user verifies the accuracy of the live tracking system")
    public void theUserVerifiesTheAccuracyOfTheLiveTrackingSystem() {
        verifyTrackingAccuracy();
    }

    @Then("live tracking information is accurate and reliable")
    public void liveTrackingInformationIsAccurateAndReliable() {
        verifyTrackingReliability();
    }
}